metasmoke.py

```diff
@@ -32,16 +32,103 @@
 
 
-MAX_FAILURES = 10  # Preservative, 10 errors = MS down
 NO_ACTIVITY_PINGS_TO_STANDBY = 8
 NO_ACTIVITY_PINGS_TO_REPORT = 4
 
 
 # noinspection PyClassHasNoInit,PyBroadException,PyUnresolvedReferences,PyProtectedMember
 class Metasmoke:
     status_pings_since_scan_activity = 0
-    last_ping_scan_time = 0
-    last_ping_posts_scanned = 0
+
+    class AutoSwitch:
+        """ Automatically switch metasmoke status """
+        # Last change was on 20 May 2020.
+        MAX_FAILURES = 10  # 10 failures == ms down
+        MAX_SUCCESSES = 1  # 1 success == ms up
+        counter = 0
+        auto = True
+        rw_lock = threading.Lock()
+
+        @staticmethod
+        def to_off():
+            """ Road to switching metasmoke down """
+            with Metasmoke.AutoSwitch.rw_lock:
+                if Metasmoke.AutoSwitch.counter < 0:
+                    # Consecutive counter. Switch sign.
+                    Metasmoke.AutoSwitch.counter = 0
+                Metasmoke.AutoSwitch.counter += 1
+                current_counter = Metasmoke.AutoSwitch.counter
+                current_auto = Metasmoke.AutoSwitch.auto
+            # MAX_FAILURES is constant so no lock.
+            if current_counter > Metasmoke.AutoSwitch.MAX_FAILURES and\
+               GlobalVars.MSStatus.is_up() and current_auto:
+                log("warning", "Last {} connections to metasmoke failed".format(Metasmoke.AutoSwitch.counter) +
+                               " Setting metasmoke status to down.")
+                chatcommunicate.tell_rooms_with("debug", "**Warning**. {}:".format(GlobalVars.location) +
+                                                         " Last {} connection to metasmoke".format(current_counter) +
+                                                         " failed. Setting metasmoke status to **down**.")
+                Metasmoke.ms_down()
+
+        @staticmethod
+        def to_on():
+            """ Road to switching metasmoke up """
+            with Metasmoke.AutoSwitch.rw_lock:
+                if Metasmoke.AutoSwitch.counter > 0:
+                    # Consecutive counter. Switch sign.
+                    Metasmoke.AutoSwitch.counter = 0
+                Metasmoke.AutoSwitch.counter -= 1
+                # Negative values for success
+                current_counter = -Metasmoke.AutoSwitch.counter
+                current_auto = Metasmoke.AutoSwitch.auto
+            # MAX_SUCCESSES is constant so no lock.
+            if current_counter > Metasmoke.AutoSwitch.MAX_SUCCESSES and\
+               GlobalVars.MSStatus.is_down() and current_auto:
+                # Why use warning? Because some action may be needed if people don't think metasmoke is up.
+                log("warning", "Last {} connections to metasmoke succeeded".format(Metasmoke.AutoSwitch.counter) +
+                               " Setting metasmoke status to up.")
+                chatcommunicate.tell_rooms_with("debug", "**Warning**. {}:".format(GlobalVars.location) +
```

Comment:
While having this be sent as a "warning" level for the `log` is fine, turning MS on shouldn't be presented as a warning in chat. Presumably, turning MS on is a good thing, which we should tell people about, but shouldn't warn them about it.